src/Build/BackEnd/Components/SdkResolution/SdkResolution.cs

```diff
@@ -0,0 +1,198 @@
+                        if (result != null && result.Success)
+                        {
+                            LogWarnings(logger, buildEventContext, importElementLocation, result);
+                            return result.Path;
+                        }
+
+                        results.Add(result);
+                    }
+                    catch (Exception e)
+                    {
+                        logger.LogFatalBuildError(buildEventContext, e, new BuildEventFileInfo(importElementLocation));
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                logger.LogFatalBuildError(buildEventContext, e, new BuildEventFileInfo(importElementLocation));
+                throw;
+            }
+
+            foreach (var result in results)
+            {
+                LogWarnings(logger, buildEventContext, importElementLocation, result);
+
+                if (result.Errors != null)
+                {
+                    foreach (var error in result.Errors)
+                    {
+                        logger.LogErrorFromText(buildEventContext, "", "", "",
+                            new BuildEventFileInfo(importElementLocation), error);
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        private void Initialize(ILoggingService logger, BuildEventContext buildEventContext, ElementLocation location)
+        {
+            lock (_lockObject)
+            {
+                if (_resolvers != null) return;
+                _resolvers = _sdkResolverLoadingStrategy.LoadResolvers(logger, buildEventContext, location);
+            }
+        }
+
+        private static void LogWarnings(ILoggingService logger, BuildEventContext bec, ElementLocation location,
+            SdkResultImpl result)
+        {
+            if (result.Warnings == null) return;
+
+            foreach (var warning in result.Warnings)
+                logger.LogWarningFromText(bec, null, null, null, new BuildEventFileInfo(location), warning);
+        }
+
+        private class SdkLoggerImpl : SdkLogger
+        {
+            private readonly BuildEventContext _buildEventContext;
+            private readonly ILoggingService _loggingService;
+
+            public SdkLoggerImpl(ILoggingService loggingService, BuildEventContext buildEventContext)
+            {
+                _loggingService = loggingService;
+                _buildEventContext = buildEventContext;
+            }
+
+            public override void LogMessage(string message, MessageImportance messageImportance = MessageImportance.Low)
+            {
+                _loggingService.LogCommentFromText(_buildEventContext, messageImportance, message);
+            }
+        }
+
+        private class SdkResultImpl : SdkResult
+        {
+            public SdkResultImpl(ReferencedSdk referencedSdk, IEnumerable<string> errors, IEnumerable<string> warnings)
```

Comment:
Don't love this ctor overload strategy, since it's hard to differentiate. But it's pretty well hidden so it's probably fine.